Training/python/reco/Training_SNNv0.py

```diff
@@ -0,0 +1,367 @@
+        self.n_gnn_layers = len(list_outputs)
+        self.n_dense_layers = self.n_dense_layers
+
+        for i in range(self.n_gnn_layers):
+            self.GNN_layers.append(MyGNNLayer(n_dim=list_n_dim[i], num_outputs=list_outputs[i], regu_rate = self.regu_rate, name='GNN_layer_{}'.format(i)))
+            self.batch_norm.append(tf.keras.layers.BatchNormalization(name='batch_normalization_{}'.format(i)))
+            self.acti_gnn.append(tf.keras.layers.Activation("tanh", name='acti_gnn_{}'.format(i)))
+            if(self.dropout_rate > 0):
+                self.dropout_gnn.append(tf.keras.layers.Dropout(self.dropout_rate ,name='dropout_gnn_{}'.format(i)))
+
+        for i in range(self.n_dense_layers-1):
+            if(self.regu_rate < 0):
+                self.dense.append(tf.keras.layers.Dense(self.n_dense_nodes, kernel_initializer="he_uniform",
+                                    bias_initializer="he_uniform", name='dense_{}'.format(i)))
+            else:
+                self.dense.append(tf.keras.layers.Dense(self.n_dense_nodes, kernel_initializer="he_uniform",
+                                bias_initializer="he_uniform", kernel_regularizer=tf.keras.regularizers.L2(l2=self.regu_rate), 
+                                bias_regularizer=tf.keras.regularizers.L2(l2=self.regu_rate), name='dense_{}'.format(i)))
+            self.dense_batch_norm.append(tf.keras.layers.BatchNormalization(name='dense_batch_normalization_{}'.format(i)))
+            self.dense_acti.append(tf.keras.layers.Activation("sigmoid", name='dense_acti{}'.format(i)))
+            if(self.dropout_rate > 0):
+                self.dropout_dense.append(tf.keras.layers.Dropout(self.dropout_rate ,name='dropout_dense_{}'.format(i)))
+
+        n_last = 4 if self.mode == "p4_dm" else 2
+        # self.dense_dm = tf.keras.layers.Dense(6, kernel_initializer="he_uniform",
+        #                         bias_initializer="he_uniform", activation="softmax", name='dense_dm')
+        # self.dense_p4 = tf.keras.layers.Dense(2, kernel_initializer="he_uniform",
+        #                         bias_initializer="he_uniform", name='dense_p4')
+        self.dense2 = tf.keras.layers.Dense(n_last, kernel_initializer="he_uniform",
+                                bias_initializer="he_uniform", name='dense2')
+
+    @tf.function
+    def call(self, xx):
+        x_mask = xx[:,:,self.map_features['pfCand_valid']]
+
+        x_em1 = self.embedding1(tf.abs(xx[:,:,self.map_features['pfCand_particleType']]))
+        x_em2 = self.embedding2(tf.abs(xx[:,:,self.map_features['pfCand_pvAssociationQuality']]))
+        x_em3 = self.embedding3(tf.abs(xx[:,:,self.map_features['pfCand_fromPV']]))
+        # x = self.normalize(xx)
+        # x = self.scale(x)
+
+        x_part1 = xx[:,:,:self.map_features['pfCand_particleType']]
+        x_part2 = xx[:,:,(self.map_features["pfCand_fromPV"]+1):]
+        x = tf.concat((x_em1,x_em2,x_em3,x_part1,x_part2),axis = 2)
```

Comment:
better to have categorical features at the end of the tensor and have `input_dim`s for them as parameters in the constructor.

Reply:
The problem with such generalization that I see, let's say you want to just cut the tail with this input_dim [:,:,:, -input_dim:], the embedding is still different for every of the categorical variables (they have different ranges), but with the current option, you can still adjust everything yourself in the Training_SNNv0.py code

Reply:
Maybe it is better to have two separate input tensors: one with normal features other with categorical. What do you think?

Reply:
Hm, that would be ok I think, in this case we would avoid slicing op. and we can initiate input ranges for embendding from yaml, if we have `variable : [ null, categorical, a1, a2], a1 and at2 will not be concidered for scaling but we can use them for init for embendding

Reply:
@kandrosov there is one problem in the current NN: matrix with distances is generated on every call -> not efficient. This part should be taken out of recurrent call of GNN_layer https://github.com/shedprog/TauMLTools/blob/d50878fa6ec06a09933ac1e85f167df04f57c96a/Training/python/reco/Training_SNNv0.py#L60-L86

If I now create one more tensor, I have 2x calculations of distance matrices

Reply:
you should append two tensors before passing them to GNN layer. i.e.
```
normal_inputs ------------>| 
cat_input1 -> emb_layer1 ->| -> concat -> GNN_layer1 -> ...
cat_input2 -> emb_layer2 ->|
```

Reply:
I see, I also understood that this transformation for evaluating if `w` is dependent on input `x` to MyGNNLayer because of the creation of `a` and `b` but I would think about alternative tensor operation not to recalculate `w` every time, the waigths de-facto are the same for all  `MyGNNLayer`

Reply:
Yes, it could be an alternative approach. But the idea that the distances between points should change between layers is the core for the SNN architecture: we start with distances in (eta, phi) space and let the NN regrouping points for each consecutive layer in order to find a better parameter space for distance measurement.